src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/EventBookingManager.java

```diff
@@ -394,6 +444,104 @@ public EventBookingDTO requestBooking(final IsaacEventPageDTO event, final Regis
+                Date reservationCloseDate = Stream.of(calendar.getTime(), event.getDate())
+                        .min(Comparator.comparing(Date::getTime))
+                        .orElseThrow(NoSuchElementException::new);
+
+                SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
+                Map<String, String> additionalEventInformation = new HashMap<>();
+                additionalEventInformation.put("reservationCloseDate", dateFormat.format(reservationCloseDate));
+
+                // attempt to book them on the event
+                if (this.hasBookingWithStatus(event.getId(), user.getId(), BookingStatus.CANCELLED)) {
+                    // if the user has previously cancelled we should let them book again.
+                    reservation = this.bookingPersistenceManager.updateBookingStatus(event.getId(), user.getId(),
+                            BookingStatus.RESERVED, additionalEventInformation);
+                } else {
+                    reservation = this.bookingPersistenceManager.createBooking(event.getId(), user.getId(), reservingUser.getId(),
+                            BookingStatus.RESERVED, additionalEventInformation);
+                }
+                reservations.add(reservation);
+
+                try {
+                    emailManager.sendTemplatedEmailToUser(user,
+                            emailManager.getEmailTemplateDTO("email-event-reservation-requested"),
+                            new ImmutableMap.Builder<String, Object>()
+                                    .put("reservingUser", getTeacherNameFromUser(reservingUser))
+                                    .put("contactUsURL", generateEventContactUsURL(event))
+                                    .put("eventURL", String.format("https://%s/eventbooking/%s",
+                                            propertiesLoader.getProperty(HOST_NAME), event.getId()))
+                                    .put("event.emailEventDetails",
+                                            event.getEmailEventDetails() == null ? "" : event.getEmailEventDetails())
+                                    .put("event", event)
+                                    .build(),
+                            EmailType.SYSTEM);
+
+                } catch (SegueDatabaseException | ContentManagerException e) {
+                    log.error(String.format("Unable to send event email (%s) to user (%s)",
+                            event.getId(), user.getEmail()), e);
+                }
+            }
+            return reservations;
+
+        } catch (EventIsFullException | SegueDatabaseException | EventGroupReservationLimitException | NullPointerException e) {
+            for (EventBookingDTO reservation : reservations) {
+                try {
+                    bookingPersistenceManager.deleteBooking(event.getId(), reservation.getUserBooked().getId());
```

Comment:
This would lose records for previously cancelled users. I think we can live with that but a database transaction would not lose that information.

Reply:
If we have facilities for defining transactions via Java, then I'm happy to consider this option. However, considering this is done after acquiring a lock, the likelihood of incurring in problems here is equivalent to that of incurring in a `SegueDatabaseException`, at which point I'm not sure that a transaction would be of much help anyway.

Reply:
We do have access to transactions from Java, we use them for user Badge management, but I don't think it's worth doing now.

Reply:
Any of EventIsFullException | SegueDatabaseException | EventGroupReservationLimitException | NullPointerException would cause the deletion of previously cancelled users not just SegueDatabaseExceptions.
I feel we should be catching something higher up Exception's inheritance hierarchy than NullPointerException. Currently we wouldn't roll back if any of the other unchecked runtime exceptions were to be thrown in the try block.

Reply:
Ok, I gave this method a full overhaul. This now uses a database transaction so we can rollback safely, and also avoids sending emails before completing all the reservations. Tests are broken and I can't figure out how to fix them though.